core/src/main/scala/org/apache/spark/sql/execution/row/RowDeleteExec.scala

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2017 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package org.apache.spark.sql.execution.row
+
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode, ExpressionCanonicalizer}
+import org.apache.spark.sql.catalyst.expressions.{Attribute, BindReferences}
+import org.apache.spark.sql.execution.{SparkPlan, TableMutationExec}
+import org.apache.spark.sql.sources.ConnectionProperties
+import org.apache.spark.sql.store.CodeGeneration
+import org.apache.spark.sql.types.{StructField, StructType}
+
+/**
+ * Generated code plan for deletes in a row table.
+ */
+case class RowDeleteExec(child: SparkPlan, resolvedName: String,
+    tableSchema: StructType, keyColumns: Seq[Attribute],
+    connProps: ConnectionProperties, onExecutor: Boolean)
+    extends TableMutationExec(forUpdate = false) {
+
+  @transient private var result: String = _
+  @transient private var stmt: String = _
+  @transient private var batchCount: String = _
+
+  override protected def doProduce(ctx: CodegenContext): String = {
+    val (initCode, commitCode, endCode) = connectionCodes(ctx)
+    result = ctx.freshName("result")
+    ctx.addMutableState("long", result, s"$result = -1L;")
+    stmt = ctx.freshName("statement")
+    batchCount = ctx.freshName("batchCount")
+    val childProduce = doChildProduce(ctx)
+    s"""
+       |if ($result >= 0L) return;
+       |$initCode
+       |try {
+       |  final java.sql.PreparedStatement $stmt = $connTerm.prepareStatement(
+       |    "${s"DELETE FROM $resolvedName WHERE $whereClauseString"}");
```

Comment:
> I don't think I understand this code. The whereClauseString is just equality comparison. Why is a prepared statement with batching needed?

Its equality comparison on only the key columns which is row-by-row updates. Given any arbitrary bulk delete/update string, the Snappy parser will transform it to a select+RowUpdateExec/DeleteExec plan where the update/delete forms row-wise updates using the primary key columns for each row as returned by the select plan (i.e. its child).

> I mean we are delegating deletes and updates to gemxd.

Changed to PK operations on gemxd. We could do direct region put/delete too if required but this was a bit easier.

> I don't see any scan happening in snappy side of the code. What is the role of generated code? Same for RowDeleteExec.

Scan is the child of this plan. SnappyParser will transform and "update ... query" into a RowUpdateExec plan  on top of query plan like usual. Same for RowDeleteExec. For column tables, there will be new implementations of these two plan in terms of what outlined in the spec.

We also need to add equivalent API support but need to discuss well first. I didn't think it important for GA so skipped it for now. Its not easy to express it in terms of DataFrameWriter -- perhaps introducing new "UpdateRow" and "DeletedRow" then a Dataset of those, and new writers like DataUpdateFrameWriter, DataDeleteFrameWriter?